[4.0] Decrease the amount of node saves

scottwulf opened the pull request:
Rebase of PR #80
Only save the node when attribute(s) have changed
Do not fetch monitor secret if we're the master

scottwulf committed:
Decrease the amount of node saves

Rebase of PR #80
Only save the node when attribute(s) have changed
Do not fetch monitor secret if we're the master

Itxaka commented on chef/cookbooks/ceph/recipes/mon.rb:
```diff
@@ -85,10 +87,11 @@
       add_key.run_command
       add_key.error!
 
+      # no need to check if the attribute is already set: it's part of the
```
comment not finished

Itxaka commented on chef/cookbooks/ceph/recipes/osd.rb:
```diff
@@ -127,7 +137,7 @@
         device["device"] = d.name
         node.set["ceph"]["osd_devices"].push(device)
-        node.save
+        dirty = true
```
We were saving here explicitly for a reason, probably becasue we are using this values `["ceph"]["osd_devices"]` immediately below.

Are we sure this is still working as before? I would say yes, as its only used in this recipe, but would like confirmation that this has been tested properly and the devices are being picked up in the recipe afterwards without saving the node.

scottwulf replied:
Yes, the array of devices is enumerable (i.e. usable) immediately below even after moving node.save to the end of the file. The node.save is only needed to persist the values for subsequent executions of the cookbook.

Itxaka replied:
Thanks!

Itxaka reviewed (dismissed):
Looks good, just a question and a nit

Itxaka reviewed (commented):
Looks good, just a question and a nit

scottwulf committed:
Decrease the amount of node saves

Fix multiline comment
Fix array initialization (was mistakenly initializing as a hash)

nicolasbock commented on chef/cookbooks/ceph/recipes/osd.rb:
```diff
@@ -71,7 +71,7 @@
     dirty = false
 
-    node.set["ceph"]["osd_devices"] ||= []
+    node.set["ceph"]["osd_devices"] = [] if node["ceph"]["osd_devices"].nil?
```
Same question as is in https://github.com/crowbar/crowbar-ceph/pull/103. Why the change?
